presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java

```diff
@@ -253,14 +279,105 @@ public Partitioning translateVariable(Function<VariableReferenceExpression, Vari
             if (argument instanceof ConstantExpression) {
                 newArguments.add(argument);
             }
-            else {
-                checkArgument(argument instanceof VariableReferenceExpression, format("Expect argument to be VariableReferenceExpression, but get %s (%s)", argument.getClass(), argument));
+            else if (argument instanceof VariableReferenceExpression) {
                 Optional<RowExpression> newArgument = translator.apply((VariableReferenceExpression) argument);
                 if (!newArgument.isPresent()) {
                     return Optional.empty();
                 }
                 newArguments.add(newArgument.get());
             }
+            else {
+                return Optional.empty();
+            }
+        }
+
+        return Optional.of(new Partitioning(handle, newArguments.build()));
+    }
+
+    // Maps VariableReferenceExpression in both partitions to an COALESCE expression, keeps constant arguments unchanged.
+    public Optional<Partitioning> translateToCoalesce(Partitioning other)
+    {
+        checkArgument(this.handle.equals(other.handle), "incompatible partitioning handles: %s != %s", this.handle, other.handle);
+        checkArgument(this.arguments.size() == other.arguments.size(), "incompatible number of partitioning arguments: %s != %s", this.arguments.size(), other.arguments.size());
+        ImmutableList.Builder<RowExpression> arguments = ImmutableList.builder();
+        for (int i = 0; i < this.arguments.size(); i++) {
+            RowExpression leftArgument = this.arguments.get(i);
+            RowExpression rightArgument = other.arguments.get(i);
+            if (leftArgument instanceof ConstantExpression) {
+                arguments.add(leftArgument);
+            }
+            else if (rightArgument instanceof ConstantExpression) {
+                arguments.add(rightArgument);
+            }
+            else if (leftArgument instanceof VariableReferenceExpression && rightArgument instanceof VariableReferenceExpression) {
+                VariableReferenceExpression leftVariable = (VariableReferenceExpression) leftArgument;
+                VariableReferenceExpression rightVariable = (VariableReferenceExpression) rightArgument;
+                checkArgument(leftVariable.getType().equals(rightVariable.getType()), "incompatible types: %s != %s", leftVariable.getType(), rightVariable.getType());
+                arguments.add(new SpecialFormExpression(COALESCE, leftVariable.getType(), ImmutableList.of(leftVariable, rightVariable)));
+            }
+            else {
+                return Optional.empty();
+            }
+        }
+        return Optional.of(new Partitioning(this.handle, arguments.build()));
+    }
+
+    public Optional<Partitioning> translateRowExpression(Map<VariableReferenceExpression, RowExpression> inputToOutputMappings, Map<VariableReferenceExpression, RowExpression> assignments, TypeProvider types)
```

Comment:
nit: I would declare the `inputToOutputMappings` as `Map<VariableReferenceExpression, RowExpression>` for more clarity

Reply:
Sorry, i meant `Map<VariableReferenceExpression, VariableReferenceExpression>`

Reply:
Nevermind, the mappings can contain constants